### PR TITLE
Chore: Walk the ast as less as possible

### LINKED ISSUE #1925: [Feature] Find a way to walk the AST only once when analyzing JS
# 🚀 Feature request

## Description

Right now when a hint needs to access the JavaScript AST  to analyze anything it needs to create a new `Linter` object:

```ts
public constructor(context: HintContext<ScriptEvents>) {
        const linter = new Linter();

        linter.defineRule('eslint-like-rule', {
            create(eslintContext) {
// ...
            },
            meta: ...
        } );
}
```

We should find a way to only create one shared `linter` object.

## What scenarios will this solve?

Perf improvements and memory reduction

### PR DESCRIPTION
**Pull request by sarvaje**
Fix #1925

## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/hint)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

~~- [ ] Added/Updated related documentation.~~
- [x] Added/Updated related tests.

### CODE REVIEW

File: packages/parser-javascript/src/parser.ts [hunk 1]
```diff
@@ -26,6 +42,40 @@ export default class JavascriptParser extends WebhintParser<ScriptEvents> {
...
+             */
+            const walkArray: WalkArray = [];
+
+            /*
+             * For now, only `walk.simple` is supported.
+             * When support for any other method is added we will need
+             * to reuse this method (in case of `ancestor`) or create
+             * a new one if the signature is not the same.
+             */
+            const simple = (node: ESTree.Node, visitors: NodeVisitor, base?: NodeVisitor, state?: any) => {
```
Comment (molant) [inline packages/parser-javascript/src/parser.ts]: Do we want to add support for the [other methods](https://github.com/acornjs/acorn/tree/master/acorn-walk#interface) while we are at it?

Comment (sarvaje) [reply packages/parser-javascript/src/parser.ts]: ok

Comment (sarvaje) [reply packages/parser-javascript/src/parser.ts]: @molant done!

File: packages/parser-javascript/src/parser.ts [hunk 2]
```diff
@@ -26,6 +42,61 @@ export default class JavascriptParser extends WebhintParser<ScriptEvents> {
             const ast = parse(sourceCode, options) as ESTree.Node;
             const tokens = [...tokenizer(sourceCode, options)];
 
+            /*
+             * We will need an array for each walk method supported.
```
Comment (molant) [inline packages/parser-javascript/src/parser.ts]: I think this comment is now obsolete

File: packages/parser-javascript/src/parser.ts [hunk 3]
```diff
@@ -34,6 +105,34 @@ export default class JavascriptParser extends WebhintParser<ScriptEvents> {
                 tokens,
                 walk
             });
+
+            Object.entries(walkArrays).forEach(([methodName, walkArray]) => {
+                walkArray.forEach(([{ node, state, base }, visitors]) => {
+                    let fullVisitors: any | Function = {};
```
Comment (molant) [inline packages/parser-javascript/src/parser.ts]: What about `allVisitors`?
```suggestion
                    let allVisitors: any | Function = {};
```

File: packages/parser-javascript/src/types.ts [hunk 4]
```diff
@@ -23,16 +23,18 @@ type NodeTypeForValue<N, T> = N extends { type: T } ? N : never;
  * Object with optional properties for each possible value of `type`.
  * Each property references a method taking a `node` of the type
  * which corresponds to the value of `type` represented by the property name.
+ * And optionaly the ancestors of that node if we are using `ancestor` for the
```
Comment (molant) [inline packages/parser-javascript/src/types.ts]:
```suggestion
 * And optionally the ancestors of that node if we are using `ancestor` for the
```

**Comment by sarvaje** [comment]
@molant done!

**Comment by sarvaje** [comment]
I have added a few comments. Let me know if I have to change anything.

File: packages/parser-javascript/src/parser.ts [hunk 5]
```diff
@@ -35,6 +165,55 @@ export default class JavascriptParser extends WebhintParser<ScriptEvents> {
                 tokens,
                 walk
             });
+
+            /**
+             * After all the hints has register their NodeVisitor or callback,
```
Comment (molant) [inline packages/parser-javascript/src/parser.ts]:
```suggestion
             * After all the hints have register their NodeVisitor or callback,
```

**Pull request closed without merging**